initial batch

ruseinov opened the pull request:
Closes #102 
This is an initial version of batch tx.
My only concern here is I create my own batch tx type to be able to pass transactions with messages down the line.

If that looks good - I'll add tests and plug that in.

ruseinov commented:
@ethanfrey this has been sitting in my git stash for a while as we're busy with other pressing things, so I figured it needs a little push as it provides us with some crucial functionality for bnsd

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,136 @@
+		return
+	}
+
+	if batchMsg, ok := msg.(*ExecuteBatchMsg); ok {
```
pp: if `!ok -> next` to have the main flow aligned on the left.

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,136 @@
+		return
+	}
+
+	if batchMsg, ok := msg.(*ExecuteBatchMsg); ok {
```
pp: same as above

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,136 @@
+		logs[i] = r.Log
+		payments += r.GasUsed
+		if len(r.Diff) > 0 {
+			diffs = append(diffs, r.Diff...)
```
q: would adding, removing the same validator work with tendermint?

ethanfrey replied:
I think we join between multiple tx before sending them to tendermint, and if the same validator id appears multiple times, we just use the last value. This reduction will work with the same validator appearing once in two distinct transactions, or twice in one transaction.

https://github.com/iov-one/weave/blob/master/app/store.go#L335-L347

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,136 @@
+			diffs = append(diffs, r.Diff...)
+		}
+		if len(r.Tags) > 0 {
+			tags = append(tags, r.Tags...)
```
q: what happens when we add `foo:bar` and `foo:other` ? I remember some discussion with @ethanfrey that this is not going to work with tendermint. We can can test and fail the batch delivery if it is not supported,IMHO.

ethanfrey replied:
This is a bit of a bug tendermint side.

There is no fail, and the tags are all stored. You can also search for them.
However, a subscription will only receive one of the tags with the same key.

We work around this in weave by placing what would be `key=value` into `key:value=1`, so they should always be unique unless it really is the same tag.

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,136 @@
+	}
+	return weave.DeliverResult{
+		Data:    amino.MustMarshalBinary(datas),
+		Log:     strings.Join(logs, "\n"),
```
q: same questions about data and log as with CheckResult

ethanfrey replied:
remove this `amino.MustMarshalBinary(datas)`.

Also most formats is fine, as long as the client can parse it out and map the results back to the original messages in the transaction. You can use something like this to hold the data: https://github.com/iov-one/weave/blob/master/app/results.proto

You could also use a similar structure for the logs, which is more correct than assuming no `"\n"` in the individual log messages. The existing join is "ok", but I would kind of prefer the protobuf approach

ruseinov replied:
@ethanfrey got it, let's default to protobuf approach everywhere to be consistent. Also, protobuf kind of solves the issue with parsing things for us.

@alpe you were right, I just didn't fully understand what you were meaning by that.

alpe commented on x/batch/msg.go:
```diff
@@ -0,0 +1,25 @@
+}
+
+func (m *ExecuteBatchMsg) Validate() error {
+	for _, v := range m.Messages {
```
req: Please limit the number of messages allowed in a batch. Also batches must not include other batches but that is probably defined in protobuf.

ethanfrey replied:
Maybe 10? But yeah, any constant is good to enforce and easily adjust later

ruseinov replied:
@ethanfrey 10 it is, going to be a constant so we can change it.
I keep wondering if it might be cool to include this config variable in genesis and default to something when it does not exist?

ethanfrey replied:
> I keep wondering if it might be cool to include this config variable in genesis and default to something when it does not exist?

Yes, nice idea. We can do this a few other places, to read constants from the genesis file. Issue is, we only get the genesis file on `init`, so for restart, this needs to be persisted to the kvstore.

Please make some reusable code for this, and happy to use this other places. I started doing this ages ago for fees, but never finished the PR as I had 1000 other things and it wasn't priority. Maybe this code gives some ideas, but it is not polished or tested fully, so don't copy, just compare those ideas to yours.
https://github.com/iov-one/weave/pull/41

ruseinov replied:
@ethanfrey I've been wondering: why don't we use initializer for that as well?
And then query the kv store for values we need.

This has it's drawbacks though, but if we need to store this in KV, why bother?

ruseinov replied:
I need to think about this a bit more though

ruseinov commented on x/batch/codec.proto:
```diff
@@ -0,0 +1,48 @@
+    repeated Msg messages = 1 [(gogoproto.nullable) = false];
+    // Msg is a workaround as protobuf does not support repeated oneof
+    message Msg {
+        oneof msg {
```
@alpe Since this is an extension I'm afraid we will have to deal with all the messages weave supports.
We are dealing with limitations of protobuf here mostly. 

It's super easy to fix this if we could forgo the protobuf codec for this particular case, but I'm afraid we have to be repetitive and verbose, look at how bnsd and bcpd codecs look for "inspiration".

ethanfrey replied:
I think we will need to repeat somehow, but alex is right, this is application specific and should be in the `cmd/*/app` dir, not in batch. If in batch, then batch is tied to exactly one app and not re-usable.

ruseinov replied:
@ethanfrey makes sense, I always forget about the fact that one could extend weave outside it's codebase and add messages.
I will code that to some BatchMsg interface to allow defining these within the app.

ruseinov commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,136 @@
+		payments += r.GasPayment
+	}
+	return weave.CheckResult{
+		Data:         amino.MustMarshalBinary(datas),
```
@alpe It has nothing to do with protobuf.
This implementation is inspired by @ethanfrey implementation from older cosmos sdk which used go-wire, which is now go-amino.

go-amino is going to be used anyway as we are dealing with tendermint, so this is a valid approach in my understanding.

ethanfrey replied:
Oh, please don't use go-amino. 🙏 
There is absolutely *zero* support for it in any language outside of golang. Even with months of requests and assurances from the tendermint team that it is easy to port.

I showed the version I wrote before to give some inspiration, but please don't use this codec..

ethanfrey replied:
BTW, tendermint doesn't use go-amino outside of the p2p interface.

The main use is in cosmos-sdk. And go-amino is exactly the reason we have no adaptor client side that works with them

ruseinov commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,136 @@
+	}
+	return weave.CheckResult{
+		Data:         amino.MustMarshalBinary(datas),
+		Log:          strings.Join(logs, "\n"),
```
This also has been inspired by earlier implementations, I think newline is good enough and also individual messages are not very helpful in this case, it is easy to break down by new line as needed though

ruseinov commented:
@ethanfrey @alpe thanks for clarifications, will do these changes today

ruseinov commented:
@ethanfrey was not sure about amino so figured I would clone your wire approach and see what options we have

ruseinov commented:
@alpe ci is showing something weird, as there is no such message anymore. gotta investigate, maybe just a travis glitch

webmaster128 commented on cmd/bcpd/app/codec.proto:
```diff
@@ -13,7 +14,20 @@ import "github.com/iov-one/weave/x/validators/codec.proto";
+}
+
+
+message Sum {
```
Is `Sum` a union type (`Sum` queals `cash.SendMsg` or `namecoin.NewTokenMsg` or ... or `BatchMsg`)? Then I think `Union` would be the better name

ruseinov replied:
It is

webmaster128 commented on cmd/bcpd/app/codec.proto:
```diff
@@ -27,13 +41,11 @@ message Tx {
     multisig.UpdateContractMsg update_contract_msg = 9;
     // validators actions
     validators.SetValidatorsMsg set_validators_msg = 10;
+    BatchMsg batch_msg = 11;
```
Is this supposed to be recursive? A BatchMsg contains multiple BatchMsg that contain multiple BatchMsg …?

ruseinov replied:
I think you are right, we need to implement a manual validation there, so I guess it's best if we separate tx and batchMsg sum stuff

webmaster128 commented on cmd/bcpd/app/codec.proto:
```diff
@@ -13,7 +14,20 @@ import "github.com/iov-one/weave/x/validators/codec.proto";
 message Tx {
   // msg is a sum type over all allowed messages on this chain.
-  oneof sum{
+  Sum sum = 1;
```
Is `sum` a good name for this field? Why not `message` anaolue to `BatchMsg`?

ruseinov replied:
It's not, it's just what we had before

ruseinov commented:
@webmaster128 nice catch, I didn't think about batch in terms of being included in sum.
The idea was to have the same list of messages for both tx and batch, but I can see now that this won't work. I will revert these changes and duplicate

ethanfrey commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,164 @@
+	}
+
+	if batchMsg, ok := msg.(Msg); ok {
+		msgList := batchMsg.MsgList()
```
No check of max length here?
 I would pull something like:

```
	msg, err := tx.GetMsg()
	if err != nil {
		return res, err
	}
 	if batchMsg, ok := msg.(Msg); ok {
		msgList := batchMsg.MsgList()
		if len(msgList) > maxBatcMessages {
			//TODO: Figure out if this is the correct error
			return res, errors.ErrTooLarge()
		}
```

into a `validate()` method that extracts the message list (or nil or error), then we have the same checks on Check and Deliver.

ethanfrey commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,164 @@
+	datas := make([][]byte, len(checks))
+	logs := make([][]byte, len(checks))
+	var allocated, payments int64
+	for i, r := range checks {
```
This seems a bit clunky coming from looking at some functional code, but is correct and probably the best we can do in golang.

ruseinov commented on cmd/bcpd/app/codec.proto:
```diff
@@ -27,13 +28,38 @@ message Tx {
   }
   // fee info, autogenerates GetFees()
-  cash.FeeInfo fees = 20;
+  cash.FeeInfo fees = 12;
```
@ethanfrey unfortunately at this point we might need to break things, I will check what's possible and maybe introduce this breaking change as part of this PR. Need to revisit all the numbers and see if they could still fit.

ruseinov replied:
@ethanfrey All good, got this fixed.

ruseinov commented on cmd/bnsd/app/examples.go:
```diff
@@ -67,41 +61,6 @@ func Examples() []commands.Example {
 	}
 	tx.Signatures = []*sigs.StdSignature{sig}
 
-	guest := crypto.GenPrivKeyEd25519().PublicKey().Address()
```
@ethanfrey this is an artifact of my revert, going to get this back

alpe commented on cmd/bcpd/app/app.go:
```diff
@@ -44,6 +45,8 @@ func Chain(minFee x.Coin, authFn x.Authenticator) app.Decorators {
 		// cannot pay for fee with hashlock...
 		hashlock.NewDecorator(),
+		// make sure we execute all the transactions in batch before the save point
+		batch.NewDecorator(),
```
req: the batch decorator should be ordered first after `utils.NewSavepoint().OnCheck()` to apply the other decorators at each message of the batch. Otherwise the context could not be set for them.

alpe commented on cmd/bcpd/app/msg.go:
```diff
@@ -0,0 +1,56 @@
+			case *BatchMsg_Union_SetValidatorsMsg:
+				return t.SetValidatorsMsg, nil
+			default:
+				return nil, errors.ErrUnknownTxType(t)
```
pp: I am afraid we will easily forget about this function when adding new message types to protobuf. Reflection may be helpful indeed. Otherwise only high level tests would cover this. But we are in the same situation with TX type already.

ruseinov replied:
@alpe I think you are right. 
@ethanfrey would you mind if I convert this and the tx switch cases to reflection instead? it is such a pain to maintain that.

ethanfrey replied:
I don't know how you intend to fix this with reflection. But if you can write and test a generic function for this in less than 2 hours, go for it. Otherwise, I don't see this as priority now, and can be made an issue for future pr.

ruseinov replied:
@ethanfrey that should be "easily" fixable via reflection, because these messages comply to the same interface and then they have 1 property each which I can get. 

I'll look into it.

alpe commented on cmd/bnsd/app/app.go:
```diff
@@ -50,6 +51,8 @@ func Chain(minFee x.Coin, authFn x.Authenticator) app.Decorators {
 		// cannot pay for fee with hashlock...
 		hashlock.NewDecorator(),
+		// make sure we execute all the transactions in batch before the save point
+		batch.NewDecorator(),
```
req: same as in bcpd: please put higher in the chain

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+type BatchTx struct {
+	weave.Tx
```
q: why do you need to store `weave.Tx`? It would be more straight forward for me to impl the Marshall/ Unmarshal here with Msg.Marshall/Unmarshal or just raising an error rather than returning something unrelated.

ruseinov replied:
we don't "store" a weave.Tx, we comply to an interface which is crucial for this whole thing to work.

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,167 @@
+
+type BatchTx struct {
+	weave.Tx
+	Msg weave.Msg
```
pp: can be lower case

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,167 @@
+		return res, err
+	}
+
+	if batchMsg, ok := msg.(Msg); ok {
```
pp: if you would invert the condition, you can have the happy path aligned to the left.

alpe commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,167 @@
+		return res, err
+	}
+
+	if batchMsg, ok := msg.(Msg); ok {
```
pp: if you would invert the condition, you can have the happy path aligned to the left.

alpe commented on x/batch/msg.go:
```diff
@@ -0,0 +1,30 @@
+
+	if len(l) > MaxBatchMessages {
+		//TODO: Figure out if this is the correct error
+		return errors.ErrTooLarge()
```
pp: let's have a custom error type that is more descriptive, like too many elements

ruseinov replied:
I checked what it returns, it's descriptive enough

alpe commented on x/batch/msg.go:
```diff
@@ -0,0 +1,30 @@
+func Validate(msg Msg) error {
+	l, err := msg.MsgList()
+	if err != nil {
+		return errors.Wrap(err)
```
👍

alpe commented on x/batch/msg.go:
```diff
@@ -0,0 +1,30 @@
+	MsgList() ([]weave.Msg, error)
+}
+
+func Validate(msg Msg) error {
```
q: why not have this on the application level where you call this from? That would also help with the `MaxBatchMessages` as this limitation can be application specific.

ruseinov commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,156 @@
+	return weave.CheckResult{
+		Data: data,
+		//TODO: Is this good enough conversion? Should be, in theory
+		Log:          string(log),
```
We already encoded log into a protobuf as discussed with ethan, same with data.

ruseinov replied:
@ethanfrey @webmaster128 can you please chip in here, some history: https://github.com/iov-one/weave/pull/188#discussion_r234531843

ruseinov replied:
@ethanfrey @alpe @webmaster128 one thing here: I think the problem with protobuf approach is that we don't get consistent log strings, as we are using protobuf in batch, but not in other messages. Hence, we don't know what to expect.

Also, log messages are not human-readable when coded in protobuf and printed.

alpe replied:
Just want to add the fact, that we are not even consistent with log type in batch results. In case there is an error, we return this as a string instead.

ruseinov commented on cmd/bnsd/app/codec.proto:
```diff
@@ -57,4 +59,37 @@ enum NftType {
+            blockchain.IssueTokenMsg issue_blockchain_nft_msg = 16;
+            ticker.IssueTokenMsg issue_ticker_nft_msg = 17;
+            bootstrap_node.IssueTokenMsg issue_bootstrap_node_nft_msg = 18;
+            BatchMsg batch_msg = 19;
```
This is outdated

ruseinov commented:
@alpe finished fixes, I think we are good to go.
I will add a ticket about reflection for later

ruseinov commented:
Needs unit tests, package-level docs and possibly an acceptance test scenario

ruseinov commented:
@alpe take a look at this one for a good example of why convey and mocks https://github.com/iov-one/weave/pull/188/commits/f8c611f0ab00be34d482932c2b318d3b51fe5cba
and that's a simple case.
When I'm finished with decorator tests it's going to be way more interesting.

ruseinov commented:
@alpe all done, no need to update acceptance as batch is disabled for bnsd.

@alpe @husio @ethanfrey please take a close look at https://github.com/iov-one/weave/pull/188/files#diff-9d671778477b5502fe5fc243fd6adbab to see the reasons to use proper testing tools as opposed to using bare minimum.

I bet if we migrated to that we could reduce our test codebase, improve readability and speed up writing tests.

I'm not sure if it's only me, but some of the old school tests in this repo take ages for me to comprehend and build upon.  
Anyhow, something to think about.

husio commented on cmd/bcpd/app/msg.go:
```diff
@@ -0,0 +1,55 @@
+	messages := make([]weave.Msg, len(msg.Messages))
+	// make sure to cover all messages defined in protobuf
+	for i, m := range msg.Messages {
+		res, err := func() (weave.Msg, error) {
```
You don't need anonymous function, you can type cast without it.

ruseinov replied:
Also it is done because there is a loop and two return params, which makes things easier if they are wrapped like this.

husio commented on cmd/bcpd/app/msg.go:
```diff
@@ -0,0 +1,55 @@
+			}
+		}()
+		if err != nil {
+			return messages, err
```
If you want to return messages together with error, it's better to create the result slice as `make([]weave.Msg, 0, len(msg.Messages))` - this way you won't have nil messages in it.

husio commented on cmd/bnsd/app/app.go:
```diff
@@ -50,6 +50,8 @@ func Chain(minFee x.Coin, authFn x.Authenticator) app.Decorators {
 		// cannot pay for fee with hashlock...
 		hashlock.NewDecorator(),
+		// make sure we execute all the transactions in batch before the save point
+		//batch.NewDecorator(),
```
Is this commented on purpose?

husio commented on cmd/bnsd/app/app_batch_test.go:
```diff
@@ -0,0 +1,146 @@
+//
+//	So(cres.Data, ShouldNotBeEmpty)
+//	return dres
+//}
```
If they are no longer valid and cannot be fixed, I think it's better to remove them. They are already in git, so you can bring them back at any point.

husio commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,166 @@
+func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx,
+	next weave.Checker) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	var err error
```
No need to declare error, it's being created the line below with `:=`

husio commented on x/batch/decorator.go:
```diff
@@ -0,0 +1,166 @@
+func (d Decorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx,
+	next weave.Deliverer) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	var err error
```
No need to declare error, it's being created with `:=` line below.

husio commented on x/batch/doc.go:
```diff
@@ -0,0 +1,14 @@
+> transactions don't hit the middleware.
+
+*/
+package batch
```
:heart:

ruseinov commented:
addressed all the issues that weren't a personal preference, this is ready to go upstream I guess

ruseinov commented:
@husio if you can take a quick peek at latest changes and approve - that'd be awesome.
I have addressed most of your comments and replied to others, gotta get this one out the door finally
